### PR TITLE
DSD-1206 truncate breadcrumb text if beyond 40 characters

### PR DESCRIPTION
**Pull request by gonuguntla**
Fixes JIRA ticket [DSD-1206](https://jira.nypl.org/browse/DSD-1206)

## This PR does the following:

-  Truncate breadcrumb text if beyond 60 characters

## How has this been tested?

- Local build of Storybook

## Accessibility concerns or updates

- If a breadcrumb text is truncated, the DS `Tooltip` component  added to the truncated text and the full text displayed in the `Tooltip` component.

### Checklist:

- [x] I have updated the Storybook documentation accordingly.
- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### Front End Review:

- [x] Review the Vercel preview deployment once it is ready.

### CODE REVIEW

File: CHANGELOG.md
Comment (EdwinGuzman) [inline CHANGELOG.md]: Nitpick: wrap `Tooltip` in back ticks `.

Comment (EdwinGuzman) [reply CHANGELOG.md]: Update: "If the breadcrumb text is truncated, the DS `Tooltip` component is used to display the full text in a tooltip when the text is hovered."

File: src/components/Breadcrumbs/Breadcrumbs.stories.mdx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.stories.mdx:42]: This is 1.4.0.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: The 60 value is a [magic number](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad) so let's change that to a constant variable.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: Both `breadcrumbText` and `truncateText` are doing something similar in the first check for the string. In `breadcrumbText`, if the title is more than 60 characters, then in `truncateText`, it also checks to see if it's less than 60 characters.

I think you can combine both functions into one function. If it's less than 60 characters, just return the text. If it's more than 60 characters, then you can shorten the text, add the ellipsis, and wrap the text in a `Tooltip`.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: Two things for the id:
1 - use string interpolation 
```
id={`tooltip-${title}`}
```

2 - don't use `title` as part of the id. The `title` is a string with spaces and that doesn't work for the id. 
[id's value must not contain whitespace (spaces, tabs etc.)](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id)

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: When adding the ellipsis, use string interpolation.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Added comments to fix some issues.

We also talked about this in person: in Storybook for the `Breadcrumbs` documentation under the "Long Titles" section, add information regarding what happens to the text when it is long. Include that it will be shorten to 60 characters, an ellipsis will be added, and that the `Tooltip` is used to see the entire text.

Also, include another example in Storybook similar to the existing one where another item has very long text. Right now the example is great but it is only for the last item in the `Breadcrumbs`. Another example with another truncated text will be great to show.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (gonuguntla) [reply src/components/Breadcrumbs/Breadcrumbs.tsx]: How to add id here?

File: CHANGELOG.md
Comment (EdwinGuzman) [inline CHANGELOG.md]: Remove this space.

File: src/components/Breadcrumbs/Breadcrumbs.stories.mdx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.stories.mdx]: "Long Titles" is the title. Your description should go below it and add more context to it. For example, "Breadcrumb text titles should be short and concise. When they are over 60 characters...." etc.

File: src/components/Breadcrumbs/Breadcrumbs.stories.mdx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.stories.mdx]: The "Grandchild" text is fine but also make this great-grandchild long text as well. Then the example will show what happens when _two_ titles are extremely long.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: This update is great.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: To add a better id instead of `text`, you can try to shorten it and remove the the white space(s). You also have other better variables to use, such as `breadcrumbsID` and `index` that is available to use when calling `breadcrumbText`.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: Update this to
```
{`${subString.substr(0, subString.lastIndexOf(" "))}...`}
```

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Some more minor updates are needed but looking good.

File: CHANGELOG.md
Comment (EdwinGuzman) [inline CHANGELOG.md]: Nitpick: instead of creating a new line for this sentence, this can just go at the end of the previous line.

**Review by EdwinGuzman — COMMENTED** [review]
This looks fine to  me. Fix the snapshot tests by running `npm test -- --updateSnapshot` and commit the updates. Wait for others to add their review.

File: CHANGELOG.md
Comment (gonuguntla) [reply CHANGELOG.md]: Ok

**Review by EdwinGuzman — APPROVED** [review]
Thanks. Wait for another review before merging.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (bigfishdesign13) [inline src/components/Breadcrumbs/Breadcrumbs.tsx:39]: This functionality should be a global function.  Move this into `utils.ts` and create a new function called `truncateText`.  The default truncate length should be 60 characters, but the new function should include a `legnth` argument that will override the default character length value.

File: src/components/Breadcrumbs/Breadcrumbs.tsx
Comment (bigfishdesign13) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: Change this value to 40.

File: src/components/Breadcrumbs/Breadcrumbs.stories.mdx
Comment (bigfishdesign13) [inline src/components/Breadcrumbs/Breadcrumbs.stories.mdx]: Update text...
```
Ideally, breadcrumb page titles should be short and concise. In cases where a breadcrumb page title is over 40 characters, the `Breadcrumb` component will truncate the title and an ellipsis will be added to the end of the truncated text.  The full title will be shown in a `Tooltip` component when hovering over the truncated page title.
```

File: CHANGELOG.md
Comment (bigfishdesign13) [inline CHANGELOG.md]: Change the character count to 40.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
Overall, this works well.  It just needs some adjustments.

File: src/components/Breadcrumbs/Breadcrumbs.stories.mdx
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.stories.mdx]: I asked for this example before because when the max length was 60 characters, the text `"Grandchild with an Exceptionally Long Name"` was _not_ cut off. But now that the max length is 40 characters, the text _does_ get cut off. So you can remove this second example as both show the same thing.

File: src/utils/utils.ts
Comment (EdwinGuzman) [inline src/utils/utils.ts:122]: Add a comment for this function describing what it does.

**Review by EdwinGuzman — COMMENTED** [review]
Minor updates. The function refactor in `utils` is a good idea.

**Review by EdwinGuzman — APPROVED** [review]
Looks good. Just update the "60" in the CHANGELOG to "40". Otherwise, everything looks okay to me but this will have to wait to be merged.

**Review by bigfishdesign13 — APPROVED** [review]
Looks good!  As Edwin mentioned, this PR should not be perged yet.  This will go into the next release.

**Pull request merged**
